src/lib/HST/Util/Selectors.hs

```diff
@@ -89,3 +94,96 @@ getPatVarName (S.PTuple _ _ _) = reportFatal
 getPatVarName (S.PList _ _) = reportFatal
   $ Message Error
   $ "Expected variable or wildcard pattern, got list pattern."
+
+-------------------------------------------------------------------------------
+-- Find Identifiers in Modules                                               --
+-------------------------------------------------------------------------------
+-- | Collects all identifiers in a module in a set.
+getIdentifiers :: S.Module a -> Set String
+getIdentifiers (S.Module _ _ _ decls) = Set.unions
+  (map getIdentifiersDecl decls)
+
+-- | Collects all identifiers in a declaration.
+getIdentifiersDecl :: S.Decl a -> Set String
+getIdentifiersDecl (S.DataDecl _ _ _ _) = Set.empty
+getIdentifiersDecl (S.FunBind _ ms)     = Set.unions
+  (map getIdentifiersMatch ms)
+getIdentifiersDecl (S.OtherDecl _ _)    = Set.empty
+
+-- | Collects all identifiers in a pattern matching rule.
+getIdentifiersMatch :: S.Match a -> Set String
+getIdentifiersMatch (S.Match _ name pats rhs binds)          = Set.unions
+  [ Set.singleton (prettyName name)
+  , Set.unions (map getIdentifiersPat pats)
+  , getIdentifiersRhs rhs
+  , maybe Set.empty getIdentifiersBinds binds
+  ]
+getIdentifiersMatch (S.InfixMatch _ pat name pats rhs binds) = Set.unions
+  [ Set.singleton (prettyName name)
+  , Set.unions (map getIdentifiersPat (pat : pats))
+  , getIdentifiersRhs rhs
+  , maybe Set.empty getIdentifiersBinds binds
+  ]
+
+-- | Collects all identifiers in a pattern.
+getIdentifiersPat :: S.Pat a -> Set String
+getIdentifiersPat (S.PVar _ name)                 = Set.singleton
+  (prettyName name)
+getIdentifiersPat (S.PInfixApp _ pat1 qname pat2) = Set.insert
+  (prettyName qname)
+  (Set.union (getIdentifiersPat pat1) (getIdentifiersPat pat2))
+getIdentifiersPat (S.PApp _ qname pats)           = Set.insert
+  (prettyName qname) (Set.unions (map getIdentifiersPat pats))
+getIdentifiersPat (S.PTuple _ _ pats)             = Set.unions
+  (map getIdentifiersPat pats)
+getIdentifiersPat (S.PParen _ pat)                = getIdentifiersPat pat
+getIdentifiersPat (S.PList _ pats)                = Set.unions
+  (map getIdentifiersPat pats)
+getIdentifiersPat (S.PWildCard _)                 = Set.empty
+
+-- | Collects all identifiers in a right-hand side of a pattern matching rule.
+getIdentifiersRhs :: S.Rhs a -> Set String
+getIdentifiersRhs (S.UnGuardedRhs _ expr) = getIdentifiersExp expr
+getIdentifiersRhs (S.GuardedRhss _ grhss) = Set.unions
+  (map getIdentifiersGRhs grhss)
+
+-- | Collects all identifiers in a guarded right-hand side.
+getIdentifiersGRhs :: S.GuardedRhs a -> Set String
+getIdentifiersGRhs (S.GuardedRhs _ cond expr) = Set.union
+  (getIdentifiersExp cond) (getIdentifiersExp expr)
+
+-- | Collects all identifiers in an expression.
+getIdentifiersExp :: S.Exp a -> Set String
+getIdentifiersExp (S.Var _ qname)            = Set.singleton (prettyName qname)
+getIdentifiersExp (S.Con _ qname)            = Set.singleton (prettyName qname)
```

Comment:
You should either 
 - clarify in the documentation that not the identifiers but pretty-printed qualified names are returned,
 - return the `QName`s instead and let the caller decide what to do with them or
 - traverse the `QName`s and actually just return the identifiers.

The `Fresh` effect does not have to worry about special, symbolic or qualified names. Thus, I'd prefer the latter solution.